README.md
# clojure-sree



## Usage

Connecting to databases in Clojure

    $ java -jar clojure-sree-0.1.0-standalone.jar [args]



Copyright © 2021 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-04-02
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-04-02
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/clojure-sree/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/clojure-sree/compare/0.1.0...0.1.1


test/clojure_sree/core_test.clj
(ns clojure-sree.core-test
  (:require [clojure.test :refer :all]
            [clojure-sree.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


project.clj
(defproject clojure-sree "0.1.0-SNAPSHOT"
  :description "Connecting to databases."
  :url "http://example.com/FIXME"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}
  :dependencies  [[org.clojure/clojure "1.10.1"]
                  [mysql/mysql-connector-java "5.1.44" ]
                  [org.clojure/java.jdbc "0.7.12"]
                  [de.ubercode.clostache/clostache "1.3.1"]
                  [org.postgresql/postgresql "42.1.3"]
                  [org.clojure/data.json "1.0.0"]
                  [clj-http "3.11.0"]
                  [doric "0.9.0"]
                  [stencil "0.5.0"]]
  :main ^:skip-aot clojure-sree.core
  :target-path "target/%s"
  :profiles {:uberjar {:aot :all
                       :jvm-opts ["-Dclojure.compiler.direct-linking=true"]}})


src/clojure_sree/mysql.clj
;; Using parameterized queries to prevent MySQL injection.
; A simple demonstration.

(ns clojure-sree.mysql
  (:require [clojure.java.jdbc :as j]))

;define the map of database connection
(def mysql-db {:dbtype "mysql"
               :dbname "hr"
               :classname "com.mysql.jdbc.Driver"
               :host "localhost"
               :subprotocol "mysql"
               :user "root"
               :password "Sreecharan1234"})

(defn call-proc
  [query]
  (j/query mysql-db [query]))

(defn call-prepared-query
  [query]
  (j/db-do-prepared mysql-db [query]))

(defn quer-proc-list
  [database]
  (str "SELECT specific_name FROM `information_schema`.`ROUTINES` WHERE  routine_schema='"
       database "';"))

;Query to list all the procedures
(mapcat vals (call-proc (quer-proc-list "hr")))

(defn proc-query
  [job-id]
  (#(str "CALL new_procedure('" % "')") job-id))

((comp call-proc proc-query) "IT_PROG")

(call-prepared-query "DROP PROCEDURE IF EXISTS new_rocedure;")

(def create-proc-query "CREATE PROCEDURE `new_rocedure`(IN last_nam VARCHAR(64)) BEGIN SELECT * FROM employees where last_name=last_nam; END;")

(call-prepared-query create-proc-query)

(defn proc-query-second
  [LAST-NAME]
  (#(str "CALL new_rocedure('" % "')") LAST-NAME))

((comp call-proc proc-query-second) "Austin")

(j/query mysql-db ["SELECT * FROM employees WHERE  job_id=?" "IT_PROG"])

src/clojure_sree/postgresql.clj
;;connecting to a postgreSQL database using Clojure.

(ns clojure-sree.postgresql
  (:require [clojure.java.jdbc :as j]
            [clojure.pprint :as printp]
            [doric.core :as doric]))

(def db
  "define the (map of) database connection."
  {:dbtype "postgresql"
   :dbname "public_lan"
   :host "localhost"
   :classname "com.postgres.jdbc.Driver"
   :subprotocol "postgres"
   :user "public_user"
   :password "Sreecharan"})

(def state-sql
  "Query to create the `state` table. Creates a table consisting of
   information about various Indian states."
  (j/create-table-ddl :state [[:state_id :serial "PRIMARY KEY"]
                              [:state_name "VARCHAR(32) UNIQUE"]
                              [:population_2019 "INTEGER"]
                              [:abrv "VARCHAR(2)"]]))

;`psql` query to drop the respective tables.
(j/db-do-commands db ["drop table if exists state cascade"
                      "drop table if exists state_details"
                      "drop table if exists state_culture"])

;execute the query to create table
(j/execute! db [state-sql])

;insert the data records into `state` table.
; ----------------------------------------------------------------------------------------
(j/insert! db :state {:state_name "Andhra Pradesh" :abrv "AP" :population_2019 53903393})

(j/insert-multi! db :state [{:state_name "Karnataka" :abrv "KA" :population_2019 67562686}
                            {:state_name "Tamil Nadu" :abrv "TN" :population_2019 77841267}
                            {:state_name "Rajasthan" :abrv "ZZ" :population_2019 81032689}])

(j/update! db :state {:abrv "RJ"} ["abrv = ?" "ZZ"])

(j/db-do-prepared db ["INSERT INTO state (state_name,abrv,population_2019) values (? ,?, ?)"
                      ["Kerala" "KL" 35699443] ["Maharashtra" "MH" 123144223]] {:multi? true})

(j/query db ["SELECT * FROM state WHERE state_name=?" "Rajasthan"])

(def state-record-data
  [[237882725 "Uttar Pradesh" "UP"] [85358965 "Madhya Pradesh" "MP"]
   [13606320 "Jammu Kashmir" "JK"] [39362732 "Telangana" "TG"]
   [30141373  "Punjab" "PJ"] [29436231 "Chhattisgarh"  "CG"]
   [28204692  "Haryana" "HR"] [11250858 "Uttarakhand" "UK"]
   [4169794 "Tripura" "TR"] [1570458 "Arunachal Pradesh" "AN"]
   [690251 "Sikkim" "SK"]])

(dorun (map #(j/insert! db :state {:state_name (nth %1 1)
                                   :population_2019 (first %1)
                                   :abrv (nth %1 2)}) state-record-data))

(j/insert-multi! db :state [{:state_name "Bihar"  :abrv "BH" :population_2019  124799926}
                            {:state_name "Assam" :abrv "AS" :population_2019 35607039}
                            {:state_name "Jharkhand" :abrv "JH" :population_2019  38593948}
                            {:state_name "West Bengal"  :abrv "WB" :population_2019 91276115}
                            {:state_name "Gujarat" :abrv "GJ" :population_2019 60439692}
                            {:state_name "Odisha" :abrv "OD" :population_2019  46356334}])

;----------------------------------------------------------------------------------------
;query to get all the records from `state` table.
(j/query db ["SELECT * FROM state"] {:result-set-fn count})

(defn id-state
  "Get the data based on 'abrv' or 'state_name' fields."
  [n]
  (if (= 2 (count n))
    (j/query db ["SELECT state_id FROM state where abrv = ?" n]
             {:row-fn :state_id :result-set-fn first})
    (j/query db ["SELECT state_id FROM state where state_name = ?" n]
             {:row-fn :state_id :result-set-fn first})))

(id-state "AP")

(id-state "Sikkim")

;query to create `state_details` table.
(j/execute! db [(j/create-table-ddl :state_details [[:capital_city_id :serial "PRIMARY KEY"]
                                                    [:state_id :int "REFERENCES state ON UPDATE CASCADE ON DELETE CASCADE"]
                                                    [:capital "VARCHAR(20) UNIQUE"]
                                                    [:area_sq_km "INTEGER"]
                                                    [:NSDP_percapita_2018_19_rs "INTEGER"]])])

;insert various data records into `state_details` table.
;-----------------------------------------------------------------------------------------------------
(def query-data [["Vizaq" 162972 151173 1] ["Bengaluru" 191791 212477 2] ["Chennai" 130058 193964 3]])

(defn load-data-details-state!
  "Load the table (`state_details`) with data records."
  [vecc]
  (j/insert-multi! db :state_details (map #(hash-map :state_id (get %1 3)
                                                     :capital (first %1)
                                                     :area_sq_km (second %1)
                                                     :NSDP_percapita_2018_19_rs (get %1 2))
                                          vecc)))

(load-data-details-state! query-data)

(j/insert! db :state_details {:state_id 4 :capital "Jaipur" :area_sq_km 342248 :NSDP_percapita_2018_19_rs 110606})

(j/insert-multi! db :state_details [{:state_id 18 :capital "Patna" :area_sq_km  94163 :NSDP_percapita_2018_19_rs 40982}
                                    {:state_id 6 :capital "Mumbai" :area_sq_km 307713 :NSDP_percapita_2018_19_rs 191736}
                                    {:state_id 5 :capital "Trivandrum" :area_sq_km 38863 :NSDP_percapita_2018_19_rs  204105}])

(j/insert! db :state_details {:state_id 19 :capital "Dispur" :area_sq_km 78438 :NSDP_percapita_2018_19_rs 82837})

(j/insert! db :state_details {:state_id 8 :capital "Bhopal" :area_sq_km 308252 :NSDP_percapita_2018_19_rs 90165})

(def query_data_2 [["Lucknow" 240928 66512 7] ["Srinagar" 55538 92347 9] ["Hyderabad" 112077 204488 10]
                   ["Raipur" 135194 92413 12] ["Dehradun" 55483  198738 14]])

(load-data-details-state! query_data_2)
;------------------------------------------------------------------------------------------

;query to get information of `state_details` table.
(j/query db ["SELECT state_id from state_details"] {:result-set-fn count})

(j/query db ["SELECT constraint_name FROM information_schema.table_constraints WHERE table_name=?" "state_details"])

;----------------------------------------------------------------------------------------------------------------------------
;query to create `state_culture` table.
(def cult-sql (j/create-table-ddl :state_culture [[:demo_id :serial "PRIMARY KEY"] [:language "VARCHAR(20)"]
                                                  [:state_animal "VARCHAR(20)"] [:state_bird "VARCHAR(20)"]
                                                  [:state_id :int "REFERENCES state UNIQUE"]]))

(j/execute! db cult-sql)

;insert data records into `state_culture` table.
(j/insert-multi! db :state_culture [{:language "Kannada" :state_animal "Elephant" :state_bird "Indian Roller" :state_id 2}
                                    {:language "Telugu" :state_animal "Blackbuck" :state_bird "Rose ringed Parakeet" :state_id 1}
                                    {:language "Tamil" :state_animal "Nilgiri tahr" :state_bird "Emerald dove" :state_id 3}
                                    {:language "Malayalam" :state_animal "Elephant" :state_bird "Great Hornbill" :state_id 5}
                                    {:language "Hindi" :state_animal "Swamp deer" :state_bird "Paradise flycatcher" :state_id 8}
                                    {:language "Hindi" :state_animal "Camel" :state_bird "Great Indian bustard" :state_id 4}
                                    {:language "Marathi" :state_animal "Giant Squirrel" :state_bird "Green Pigeon" :state_id 6}
                                    {:language "Odia" :state_animal  "Sambar" :state_bird "Indian Roller" :state_id 23}])

(j/insert! db :state_culture {:language "Assamese" :state_animal "1-horned rhinoceros" :state_bird "White wood duck" :state_id 19})

(def culture-data (j/query db ["SELECT state_bird FROM state_culture"]))
;-------------------------------------------------------------------------------------------------------------

(printp/pprint (map (comp clojure.string/upper-case #(first (vals %))) culture-data))

(printp/pprint (j/get-by-id db :state_culture 3 :demo_id))

(printp/pprint (j/get-by-id db :state 12 :state_id))

(def raw-query "SELECT a.Nsdp_percapita_2018_19_rs,b.population_2019 from state_details a inner join state b
                using (state_id) where a.State_id=?")

(defn popul-gdp-id
  "Get the details based on state id."
  [id]
  (when-let [[x] (j/query db [raw-query id])]
    x))

(popul-gdp-id 2)

(printp/pprint (map #(:state_bird %) (j/query db ["SELECT a.abrv,b.state_bird,c.capital FROM state a inner join
 state_culture b USING (state_id) inner join state_details c USING (state_id)"])))

(print (doric/table (j/query db ["SELECT s.nsdp_percapita_2018_19_rs,a.state_name FROM state_details s inner join state a
using (state_id) order by s.nsdp_percapita_2018_19_rs desc limit 5"])))

(printp/pprint (j/query db ["EXPLAIN SELECT * FROM state WHERE state_name in ('Tamil Nadu', 'Andhra Pradesh','Rajasthan') "]))

(println (doric/table (j/query db ["SELECT s.nsdp_percapita_2018_19_rs,a.state_name,b.language FROM state_details s
inner join state a using (state_id) inner join state_culture b using (state_id) order by s.nsdp_percapita_2018_19_rs desc"])))

(println (doric/table (j/query db ["SELECT x.Area_sq_km,y.state_name FROM state_details x inner join state y
 using (state_id) order by x.Area_sq_km desc"])))

(printp/print-table (map #(select-keys % [:abrv :population_2019])
                         (j/query db ["SELECT * FROM state order by population_2019 desc limit 5 offset 2"])))

(def state-area (j/query db ["SELECT a.area_sq_km,b.state_name FROM state_details a
inner join state b using (state_id) order by a.area_sq_km desc"]))

(doseq [state-pop (map #(str "Area of " (:state_name %1) " is " (:area_sq_km %1) " sq. km.") state-area)]
  (println state-pop))

(println (doric/table (j/query db ["SELECT * FROM state"])))

(println (doric/table (j/query db ["SELECT * FROM state_details"])))

(println (doric/table (j/query db ["SELECT * FROM state_culture"])))

(println (doric/table (for [details_state_full (j/query db ["SELECT a.*, b.*, c.* from state a inner join state_details b
  using (state_id) inner join state_culture c using (state_id)"])]
                        details_state_full)))

src/clojure_sree/parse_sparql.clj
;; parse a given SPARQL query, resulting in a hash-map.
;; information to be utilized in later analysis and extraction.

(ns clojure-sree.parse-sparql
  (:require [clojure.string :as s]
            [clojure.pprint :as pretty]))

(def demo-query
  "PREFIX foaf:  <http://xmlns.com/foaf/0.1/> .
  SELECT ?first_name ?last_name
  FROM nutty_graph
  WHERE {
  subject:Fred foaf:lastName ?last_name ;
  foaf:firstName ?first_name .}
  ORDER BY ?last_name")

(defn nds-wth-if
  "Check if the token ends with a full-stop, and remove it if found."
  [token]
  (if (s/ends-with? token ".")
    (apply str (drop-last token))
    token)) 

(defn smpl-sntnc
  "Parse the original query to a simple sentence."
  [query]
  (->> query
       s/split-lines
       (map s/trim)
       (s/join #" ")  
       nds-wth-if))

(defn form
  "Get the basic SPARQL form of the query."
  [query]
  (when-let [form (->> query
                       s/upper-case
                       (re-seq #"\w+")
                       (some #{"CONSTRUCT" "SELECT" "ASK" "DESCRIBE"}))]
    form))

(defn variables
  "Get variables of the query."
  [query]
  (letfn [(md-strng [strng] (re-find #"(?<=(?i)ASK|CONSTRUCT|SELECT|DESCRIBE).*?(?=\{)" strng))
          (rslt-fn [strng] (->> strng
                                (re-seq #"(?<=\?)\w+")
                                (into (sorted-set))))]
    (when-let [md-strngd (md-strng query)]
      (if (re-find #" \*" md-strngd)
        (rslt-fn query)
        (rslt-fn md-strngd)))))

(defn prefixes
  "Get the prefixes of the query as hash-map."
  [query]
  (let [prfx (re-seq #"(?<=(?i)prefix).*?>" query)
        pr (mapcat #(re-seq #"(?<=\s).*?(?=:\s)" %) prfx)
        lnk (mapcat #(re-seq #"[^<]+(?=>)" %) prfx)
        rslt (map #(vector (keyword (s/trim %1)) %2) pr lnk)]
    (into (sorted-map) rslt)))

(defn graphs
  "Get the source graph from the query."
  [query]
  (if-let [frms (re-seq #"(?<=(?i)FROM|FROM NAMED)\S*(?:\s\S+)?" query)]
    (->> frms
         (map s/trim)
         (remove #{"NAMED"})
         (into #{}))
    "Default Reference."))

(defn orderby
  "Get the order by argument from the query."
  [query]
  (let [lmt-rslt (re-find #"(?i)limit.*" query)
        lmt-st (into #{} (when (some? lmt-rslt) (s/split lmt-rslt #" ")))]
    (some->> query
             (re-find #"(?<=(?i)order by).*")
             (re-seq #"\b(?i)(?!asc|desc\b)\w+")
             (remove lmt-st)
             (into (sorted-set)))))

(defn lookuptriple
  "A generic function to get triple-lookup-pattern from where part of the string."
  [query]
  (letfn [(trpld-str [str-frst] (let [trpl (some->> str-frst
                                                    (re-find #"(?<=\{).*(?=\.)")
                                                    s/trim)
                                      strtr (when-not (nil? trpl) (-> #"^\S*\s+\S+"
                                                                      (re-find trpl)
                                                                      (s/split #" ")
                                                                      first))
                                      rplc (fn [re str] ((comp #(s/replace % ";" " ")
                                                               #(re-find (re-pattern re) %)) str))
                                      trpld (when-not (and (nil? strtr) (nil? trpl))
                                              (if (.contains trpl ";")
                                                (str (rplc ".*;" trpl) ". "
                                                     strtr
                                                     (rplc ";.*" trpl))
                                                trpl))]
                                  (some-> trpld
                                          (s/split #"\.\s"))))
          (whr-rslt [strng] (->> strng
                                 trpld-str
                                 (filter #(re-find #"^(?!BIND|OPTIONAL|FILTER)" %))
                                 (map s/trim)
                                 (map #(s/split % #" "))
                                 (map #(remove #{"" "."} %))))
          (mr-thn-thr-fn [lst] (if (> (count lst) 3)
                                 (let [[frst & rst] lst
                                       cnj (->> rst
                                                (partition 2)
                                                (map #(conj % frst)))] cnj)
                                 lst))
          (fll-whr-rslt [pr-rslt] (->> pr-rslt
                                       whr-rslt
                                       (map mr-thn-thr-fn)))
          (whr-trpld [trp] (zipmap [:subject :predicate :object] (mapv nds-wth-if trp)))
          (rslt-fn [whr-tr] (if (and (= 3 (count whr-tr)) (every? string? whr-tr))
                              (whr-trpld whr-tr)
                              (map whr-trpld whr-tr)))] (->> query
                                                             fll-whr-rslt
                                                             (map rslt-fn)
                                                             flatten
                                                             distinct
                                                             vec)))

(def parse-keys
  "Vector of all the to-be-parsed keys."
  [:query-type :variables :reference-graphs :prefixes :lookup-pattern :order-value])

(def vals-fn
  "Vector of all the to-be-parsed functions."
  [form variables graphs prefixes lookuptriple orderby])

(defn parsed-query
  "Return a map of parsed information about the query.
  Using REGEXP wherever needed.
  Works for SELECT/ASk form accurately."
  [query]
  (->> vals-fn
       (map #(% (smpl-sntnc query)))
       (map hash-map parse-keys)
       (filter #(every? (comp not empty?) (vals %)))
       (into {}))) 

(pretty/pprint (parsed-query demo-query))

(defn question-mark-validator
  [query]
  (let [ptt-qs (re-pattern "(?i)(?<=select) .* (?=from)|(?i)(?<=select) .* (?=where)")
        snt-slct (re-find ptt-qs query)
        ntr-slct (when-not (nil? snt-slct) (re-find #"^[^\*]+$" snt-slct))
        bl-slct (if (some? ntr-slct) (s/trim ntr-slct) snt-slct)
        lst-slct (filterv #(not= "as" %) (s/split bl-slct #" "))
        mp-slct (map-indexed #(array-map %2 (s/starts-with? %2 "?")) lst-slct)
        slct? (every? true? (mapcat vals mp-slct))
        fnl-rslt (mapcat #(when (false? (first (vals %))) (keys %)) mp-slct)]
    (when-not slct?
      {:exception "Question mark is mandatory for variables in a SPARQL query"
       :security "Lexical error encountered."
       :fails fnl-rslt
       :count (count fnl-rslt)})))

((comp question-mark-validator smpl-sntnc) demo-query)

;; sample_result:
;{:query-type "SELECT"
;  :variables #{"avgHeight" "avgWeight" "countryCode" "height" "weight"}
;  :reference-graph "Default Reference."
;  :prefixes
;  {:dbo "http://dbpedia.org/ontology/"
;   :foaf "http://xmlns.com/foaf/0.1/"
;   :rdf "http://www.w3.org/1999/02/22-rdf-syntax-ns#"
;   :rdfs "http://www.w3.org/2000/01/rdf-schema#"
;   :walls "http://wallscope.co.uk/ontology/olympics/"}
;  :lookup-pattern
;  [{:subject "?noc", :predicate "dbo:ground", :object "?team"}
;   {:subject "?noc", :predicate "rdfs:label", :object "?countryCode"}
;   {:subject "?athlete", :predicate "rdf:type", :object "foaf:Person"}
;   {:subject "?athlete", :predicate "dbo:team", :object "?team"}
;   {:subject "?athlete", :predicate "dbo:height", :object "?height"}
;   {:subject "?athlete", :predicate "dbo:weight", :object "?weight"}]
;  :order-value #{"avgHeight"}}


src/clojure_sree/sparql.clj
(ns clojure-sree.sparql
  (:require [clojure.string :as s]
            [clj-http.client :as http]
            [clojure.data.json :as json]
            [clojure.pprint :as pretty]
            [clostache.parser :as p]
            [clojure.test :as t]))

(defn query-sparql
  ([endpoint query]
   (query-sparql endpoint query {}))
  ([endpoint query http-req]
   {:pre [(re-find #"http" endpoint)
          (re-find #"(?i)ASK|SELECT|CONSTRUCT|DESCRIBE" query)]}
   (let [response (http/get endpoint (merge (merge {:cookie-policy :standard} http-req)
                                            {:query-params {:query query}}))]
     (if (<= 200 (:status response) 299)
       (:body response)
       (throw
         (Error.
           (str "HTTP Error "
                (:status response)
                (:reason-phrase response)
                (:body response))))))))

(defn select-sparql
  "Function to attach the query to a given endpoint."
  ([endpoint query]
   (select-sparql endpoint query {}))
  ([endpoint query http-req]
   {:pre [(re-find #"http"  endpoint)
          (string? query)
          (re-find #"(?i)SELECT" query)]}
   (-> (query-sparql endpoint query (merge http-req {:accept "application/sparql-results+json"}))
       (json/read-str)
       (get "results")
       (get "bindings"))))

(def endpoint "http://localhost:3030/dataolymp/sparql")

(defn convert-to-decimal
  [token]
  (try
    (if (re-matches #"^[\d+\.]+$" token)
      (->> token
           (Double/parseDouble)
           (format "%.2f"))
      token)
    (catch Exception _)))

(defn mapfn
  "Get the result as a map from a response to the query request."
  [keysfn map-val]
  (letfn [(gt-key [map-param-key] (->> map-param-key
                                       (conj '("value"))
                                       (get-in map-val)
                                       convert-to-decimal))]
    (zipmap keysfn (map gt-key keysfn))))

(defn getnrecords
  "Get `n` records from the query."
  [n result]
  (let [keyfn ((comp keys first) result)
        mapfn (partial mapfn keyfn)]
    (->> result
         (map mapfn)
         (take n))))

; Illustrations:
; Firstly, load the `olympics` RDF triple store to the local jena fuseki server.

;----------------------------------------------------------------------------------
;Query to list each country alongside the average height and weight of its athletes

(def height-query
  "PREFIX walls: <http://wallscope.co.uk/ontology/olympics/>
   PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
   PREFIX dbo: <http://dbpedia.org/ontology/>
   PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
   PREFIX foaf: <http://xmlns.com/foaf/0.1/>
   SELECT ?countryCode
  (AVG(?height) As ?avgHeight)
  (AVG(?weight) As ?avgWeight)
  WHERE {
  ?noc dbo:ground ?team ;
       rdfs:label ?countryCode .
  ?athlete rdf:type   foaf:Person ;
            dbo:team   ?team ;
            dbo:height ?height ;
            dbo:weight ?weight .}
  GROUP BY ?countryCode
  ORDER BY DESC(?avgHeight)")

(def result-height (select-sparql endpoint height-query))

(pretty/print-table (getnrecords 4 result-height))

;----------------------------------------------------------------------------------------------
;Query to list an year alongside the oldest Judo competitor in that year.

(def judo-oldest-query
  "PREFIX walls: <http://wallscope.co.uk/ontology/olympics/>
   PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
   PREFIX foaf: <http://xmlns.com/foaf/0.1/>
   PREFIX dbp: <http://dbpedia.org/property/>
   SELECT ?year (MAX(?age) As ?maxAge)
   WHERE {
       ?instance walls:games   ?games ;
                 walls:event   ?event ;
                 walls:athlete ?athlete .
       ?event rdfs:subClassOf <http://wallscope.co.uk/resource/olympics/sport/Judo> .
                 ?games dbp:year ?year .
                 ?athlete foaf:age ?age .}
   GROUP BY ?year
   ORDER BY ?year")

(def judo-result (select-sparql endpoint judo-oldest-query))

(pretty/print-table (getnrecords 4 judo-result))

;---------------------------------------------------------------------------------
; Query to list names of every athlete with at least one medal, alongside their total number of medals

(def query
  "PREFIX walls: <http://wallscope.co.uk/ontology/olympics/>
   PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
   SELECT ?name (COUNT(?athlete) As ?noOfMedals)
   WHERE {
   ?instance walls:athlete ?athlete ;
             walls:medal   ?medal .
   ?athlete  rdfs:label    ?name .}
   GROUP BY ?name
   ORDER BY DESC(?noOfMedals)")

(def result-medal (select-sparql endpoint query))

;; return details of only top 3 players from the lazy sequence.

(pretty/print-table (getnrecords 3 result-medal))

; ----------------------------------------------------------------------------------------------------------------

;; For higher expressiveness in generating SPARQL, use a templating language `clostache`.

(def stored-procedure-limit
  "PREFIX walls: <http://wallscope.co.uk/ontology/olympics/>
  PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
  SELECT ?name (COUNT(?athlete) As ?noOfMedals)
   WHERE {
   ?instance walls:athlete ?athlete ;
            walls:medal   ?medal .
   ?athlete  rdfs:label    ?name .}
  GROUP BY ?name
  ORDER BY DESC(?noOfMedals) LIMIT {{limit}} OFFSET {{offset}}" )

(defn render-stored-mitigate-numeric 
  "A stored procedure to find out the player name and his/her total gold medals.
  Render the stored procedure iff the entered value is a positive integer."
  [limit offset]
  (if (and (int? offset) (pos-int? limit) (>= offset 0))
    (let [rndrd-qry (p/render stored-procedure-limit {:limit limit :offset offset})
          rndrd-rslt (select-sparql endpoint rndrd-qry)
          k-rslt ((comp keys first) rndrd-rslt)
          mapfn-s (partial mapfn k-rslt)]
      (mapv mapfn-s rndrd-rslt))
    "Enter a positive integer for the parameters."))

(pretty/print-table (render-stored-mitigate-numeric 2 4))

(def query-player-list
  "PREFIX walls: <http://wallscope.co.uk/ontology/olympics/>
  PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
  SELECT DISTINCT ?string
  WHERE {
  ?instance walls:athlete ?athlete ;
            walls:medal   ?medal .
  ?athlete  rdfs:label    ?name .
   BIND( STR(?name) AS ?string )}" )

(def player-list
  "Query to get the total player list."
  (map #(get-in % ["string" "value"]) (select-sparql endpoint query-player-list)))

(def stored-procedure-name
  "PREFIX walls: <http://wallscope.co.uk/ontology/olympics/>
  PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
  SELECT ?string (COUNT(?athlete) As ?noOfMedals)
  WHERE
  {?instance walls:athlete ?athlete ;
            walls:medal   ?medal .
  ?athlete  rdfs:label    ?name .
   BIND (STR(?name) AS ?string ).}
   GROUP BY ?string
   HAVING (?string = {{ name }} )" )

(defn execute-stored-procedure
  "Execute the stored procedure to fetch the total number of Gold Medal.
  Given the player name, how many medals he/she has won."
  [query name]
  (some->>
    {:name (str "'" name "'")}
    (p/render query)
    (select-sparql endpoint)
    (map #(get-in % ["noOfMedals" "value"]))
    first
    (re-find #"\d+")
    (Integer/parseInt)))

(defn mitigating-special-character
"Stored procedure with mitigating injection attacks.
 A generic function to deal with the issue."
    [query-m parameter]
  (let [patt (re-pattern "[|!#$&*%*{^}]")
        char-spec (re-seq patt parameter)
        bool-val (empty? char-spec)
        count-char (count char-spec)]
    (if-not bool-val
      {:exception "The parameter should not contain special characters."
       :security "injection-attack-mitigation:special-char"
       :fails char-spec
       :count count-char}
      (if (some #(= parameter %) player-list)
        (execute-stored-procedure query-m parameter)
        "Player name not found."))))

;return the total gold medlas of a random player from the player-list,
; after passing through mitigation check.
(mitigating-special-character stored-procedure-name (rand-nth player-list))

(def not-allowed-words
  "A set of all the words that are part of SPARQL code but should not be there in the parameter values."
  #{"SELECT" "DELETE" "WHERE" "INSERT" "GROUP" "HAVING" "ORDER" "FROM" "OPTIONAL" "FILTER"})

(defn mitigate-keyword
  [query parameter]
  (let [tru-val (some not-allowed-words (s/split (s/upper-case parameter) #" "))]
    (if-not (empty? tru-val)
      {:exception "SPARQL script should not be part of variable values."
       :security "injection-attack-mitigation:sparql-code"}
      (if (some #(= parameter %) player-list)
        (execute-stored-procedure query parameter)
        "Player name not found."))))

;return the total gold medlas of a random player from the player-list,
;after passing through mitigation check.
(mitigate-keyword stored-procedure-name (rand-nth player-list))

;; unit tests
(t/deftest testing-mitigate-keyword-stored-procedure

  (t/testing "with SELECT"
    (t/is (= "SPARQL script should not be part of variable values."
             (let [query-body (mitigate-keyword stored-procedure-name "Nikolay Yefim ovich Andrianov SELECT * FROM")]
               (:exception query-body)))))

  (t/testing "with special characters"
    (t/is (= 2 (let [query-body (mitigating-special-character stored-procedure-name "Rudolph Ludewyk *& Lewis" )]
                 (:count query-body)))))
  )

(t/run-tests)
